### PR TITLE
Add "Create desktop shortcut" feature.

### PR DESCRIPTION
**Pull request by slendidev**
This PR adds a way of adding a shortcut to directly start an instance as
requested by issues #1088 and #541. 

It is important to note, however, that this feature has not been implemented
for Mac.

Right now, I need someone that is able to test whether they can get this to
build and run under Windows.

### CODE REVIEW

**Comment by samiscool51** [comment]
managed to get it to build and run but unfortunately the option to create a shortcut doesn't appear in the instance toolbar:
image
i built it for QT5 on windows 10 LTSC 1809, i'll give QT6 a go to see if it appears there.

**Comment by samiscool51** [comment]
update: QT6 version also doesn't have the option sadly.

**Comment by glowiak** [comment]
Great. For *UNIX you just need to write a .desktop file and place it in ~/.local/share/applications so this would be much easier

**Comment by slendidev** [comment]
> managed to get it to build and run but unfortunately the option to create a shortcut doesn't appear in the instance toolbar: image i built it for QT5 on windows 10 LTSC 1809, i'll give QT6 a go to see if it appears there.

This is incredibly weird behaviour, as the option should appear only if not compiled for Macs. Are you sure you compiled correctly? Maybe you forgot to switch branch?

**Comment by slendidev** [comment]
I have built it on Windows, the option appears but it seems to not do anything.

**Comment by slendidev** [comment]
I have fixed the issues with the feature on Windows.

**Comment by samiscool51** [comment]
>Maybe you forgot to switch branch?

my bad, i forgor to switch branches. tested it again with the correct branch and it launches and works properly.

**Comment by ghost** [comment]
I'm not big into the idea of including direct windows API calls in MainWindow.cpp, but this looks pretty cool

**Comment by slendidev** [comment]
> I'm not big into the idea of including direct windows API calls in MainWindow.cpp, but this looks pretty cool

Me neither, but I don't think anyone wants to spend time re-implementing
system libraries.

**Comment by ghost** [comment]
> > I'm not big into the idea of including direct windows API calls in MainWindow.cpp, but this looks pretty cool
> 
> Me neither, but I don't think anyone wants to spend time re-implementing
> system libraries.

I think the rest of the codebase has some weird `sys_{windows,Unix,etc}.cpp` thing that implements platform-specific functions like getting CPU/RAM information. I'd recommend looking into that; you can define a universal `createInstanceShortcut` function in `sys.h` or whatever it is and then implement it for each OS in the sys files.

edit: Actually that's part of the systeminfo library so it won't work. I *would* still prefer having one universal function and individually implementing it for each OS though, in a separate file perhaps.

**Comment by slendidev** [comment]
Personally, I would leave it there. If somehow there is a need more system
libraries, then yes, those functions deserve their own special file. I don't
really see the point in making both a C++ source file and header just for one
function.

**Comment by ghost** [comment]
> Personally, I would leave it there. If somehow there is a need more system
> libraries, then yes, those functions deserve their own special file. I don't
> really see the point in making both a C++ source file and header just for one
> function.
> 

The point I see is for future potential implementations that need system specific libraries.

However you could be a gigachad and force whoever needs that to do it themselves

File: launcher/ui/MainWindow.cpp [hunk 1]
```diff
@@ -1861,6 +1901,36 @@ void MainWindow::on_actionChangeInstGroup_triggered()
...
+        return;
+  
+    auto desktop = QStandardPaths::writableLocation(QStandardPaths::DesktopLocation);
+    auto executable_path = APPLICATION->applicationFilePath();
+    auto instId = m_selectedInstance->id();
+    auto icon = APPLICATION->windowIcon();
+    auto name = m_selectedInstance->name();
+#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__NT__)
+    createLink(
+        const_cast<wchar_t *>(executable_path.toStdWString().c_str()),
```
Comment (ghost) [inline launcher/ui/MainWindow.cpp]: Any reason for the const_casts here? Why not make the LPCWSTR arguments const refs?

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
I'd leave this as a draft until it gets implemented and tested on MacOS

**Comment by slendidev** [comment]
> I'd leave this as a draft until it gets implemented and tested on MacOS

I will not be implementing this for Mac, as I have zero familiarity with it's internals. If someone wants to add support for it, a new PR can always be made.

**Comment by ghost** [comment]
> > I'd leave this as a draft until it gets implemented and tested on MacOS
> 
> I will not be implementing this for Mac, as I have zero familiarity with it's internals. If someone wants to add support for it, a new PR can always be made.

zased?

**Comment by LennyMcLennington** [comment]
Does it want Qt::endl? instead of just endl? Builds failing on qt6.

**Comment by LennyMcLennington** [comment]
Does this work with flatpak and appimage? I don't really see how it could work with flatpak, and if it won't work with them then it should at least be disabled in the UI.

Also, when you add the option to the instance toolbar, you're not checking whether the target is MacOS so the action will still be added, and when the MacOS user clicks it, it will create a .desktop file since in the actual function you only check for windows and then just #else and make a .desktop file

**Pull request merged**